### PR TITLE
Support None for the in_channels argument in ConvolutionND/DeconvolutionND

### LINKED ISSUE #2937: ConvolutionND can't  in_channels = None
`L.Convolution2D` can `input_channels = None` or omit input_channels.
But, `L.ConvolutionND` can't. 
What about making this feature be able to use in `L.ConvolutionND`.

### PR DESCRIPTION
**Pull request by mitmul**
This PR resolves #2937

### CODE REVIEW

**Comment by hvy** [comment]
@mitmul It hit me that `DeconvolutionND` should be addressed too but maybe you were on it already!

**Comment by mitmul** [comment]
Right, thanks for noting. I'll work on that too.

**Comment by mitmul** [comment]
Jenkins!! Test this please

**Comment by hvy** [comment]
Do you think we should update the docs?

**Comment by hvy** [comment]
Maybe this can be addressed in a separate PR but I think we move the hard coded test fixtures to parametrized values, and then add parameters for e.g. testing convolution ND with ndim >= 3, other kernel sizes, strides, in/out channels, etc.

File: tests/chainer_tests/links_tests/connection_tests/test_deconvolution_nd.py [hunk 1]
```diff
@@ -20,12 +20,14 @@
     'dtype': [numpy.float32],
     'use_cudnn': ['always', 'auto', 'never'],
     'used_outsize': ['case1', 'case2', 'None'],
+    'in_channels_none': [True, False],
```
[the remark this answers is not in the document]
Comment (mitmul) [reply tests/chainer_tests/links_tests/connection_tests/test_deconvolution_nd.py]: I'll check the reason and will comment here.

[the remark this answers is not in the document]
Comment (mitmul) [reply tests/chainer_tests/links_tests/connection_tests/test_deconvolution_nd.py]: @hvy Hm, I couldn't remember why it should be as is. So I fixed it.

**Comment by mitmul** [comment]
@hvy Could you take a look into it again? I just followed your suggestion.

**Comment by hvy** [comment]
Thanks for the quick fix. I'm thinking if we shouldn't support omitting the `in_channels` as in `Convolution2D` altogether, not just setting it to `None`. I think this would require us to set `ksize=None` by default and then basically copy this code https://github.com/chainer/chainer/blob/master/chainer/links/connection/convolution_2d.py#L126-L127 . Do you see any concerns doing this? (I'm wondering why the default `ksize` value for ND was different from 2D).

**Comment by mitmul** [comment]
@hvy Thank you for the suggestion. FOr Convolution/DeconvolutionND, it takes `ndim` argument as the first argument to the constructor, so that I think there's no way to omit only the second argument `in_channels`. We can give `None` to it, but can't completely omit it because it creates ambiguity, e.g., `(ndim, out_channels)` can't be discriminated from `(ndim, in_channels)` which should raise an error for insufficient arguments. So, I think it's safe not to provide the omitting option for `in_channel` argument in ND operators. Does it make sense?

**Comment by mitmul** [comment]
Jenkins, test this please

**Comment by hvy** [comment]
Shouldn't `(ndim, in_channels)` and `(ndim, out_channels)` both raise errors anyway, meaning we don't have to discriminate between those cases? 

We require at least three arguments, and if three are given, interpret them as `(ndim, out_channels, ksize)` and in case of four `(ndim, in_channels, out_channels, ksize)` . Please correct me if I'm wrong.

**Comment by mitmul** [comment]
Ah, you are right. Sorry I was confused. Thanks for clarifying it!

**Comment by mitmul** [comment]
@hvy Updated the code and add examples to the docstrings. Please check it again :)

File: chainer/links/connection/convolution_nd.py [hunk 2]
```diff
@@ -57,11 +57,15 @@ class ConvolutionND(link.Link):
 
     """
 
-    def __init__(self, ndim, in_channels, out_channels, ksize, stride=1, pad=0,
-                 nobias=False, initialW=None, initial_bias=None,
+    def __init__(self, ndim, in_channels, out_channels, ksize=None, stride=1,
+                 pad=0, obias=False, initialW=None, initial_bias=None,
```
Comment (hvy) [inline chainer/links/connection/convolution_nd.py]: Ops, `obias` a careless?

Comment (mitmul) [reply chainer/links/connection/convolution_nd.py]: @hvy Thanks, I fixed this point in the latest commit already. Please check it again :)

**Comment by mitmul** [comment]
There was a flake8 error, but fixed.

**Comment by mitmul** [comment]
Jenkins, test this please.

File: chainer/links/connection/convolution_nd.py [hunk 3]
```diff
@@ -82,7 +133,11 @@ def __init__(self, ndim, in_channels, out_channels, ksize, stride=1, pad=0,
                 initial_bias = initializers._get_initializer(initial_bias)
                 self.b = variable.Parameter(initial_bias, out_channels)
 
-    def __call__(self, x):
+    def _initialize_params(self, in_channels):
+        W_shape = (self.out_channels, in_channels) + self.ksize
```
Comment (hvy) [inline chainer/links/connection/convolution_nd.py]: Do you have to take `self.groups` into account, as it's done in `Convolution2D`?

File: chainer/links/connection/deconvolution_nd.py [hunk 4]
```diff
@@ -79,7 +131,13 @@ def __init__(self, ndim, in_channels, out_channels, ksize, stride=1, pad=0,
                 initial_bias = initializers._get_initializer(initial_bias)
                 self.b = variable.Parameter(initial_bias, out_channels)
 
-    def __call__(self, x):
+    def _initialize_params(self, in_channels):
+        W_shape = (in_channels, self.out_channels) + self.ksize
```
Comment (hvy) [inline chainer/links/connection/deconvolution_nd.py]: Ditto regarding groups!

**Comment by hvy** [comment]
Sorry for the delayed comments. Do we have to take the `groups` arguments into account when computing the size of the `W`? It seems we didn't before but was the intended behavior?

**Comment by mitmul** [comment]
@hvy Sorry for my late reply. I fixed the initialization part to consider the `groups` arguments. Could you check the code again?

**Comment by hvy** [comment]
Rather sorry for my delay. Thanks for the changes, I'll take a look!

**Comment by hvy** [comment]
Sorry again, the `groups` argument was maybe not necessary in this PR. Or rather, now I'm feeling we're responsible for adding tests for that argument too. @mitmul what do you think?

**Comment by mitmul** [comment]
Jenkins, test this please

File: chainer/links/connection/convolution_nd.py [hunk 5]
```diff
@@ -55,24 +57,73 @@ class ConvolutionND(link.Link):
...
+
+    def __init__(self, ndim, in_channels, out_channels, ksize=None, stride=1,
+                 pad=0, nobias=False, initialW=None, initial_bias=None,
                  cover_all=False, dilate=1, groups=1):
         super(ConvolutionND, self).__init__()
 
-        ksize = conv_nd.as_tuple(ksize, ndim)
+        if ksize is None:
+            ndim, out_channels, ksize, in_channels = \
+                ndim, in_channels, out_channels, None
```
Comment (hvy) [inline chainer/links/connection/convolution_nd.py]: Is the `ndim` reassignment redundant?

File: chainer/links/connection/deconvolution_nd.py [hunk 6]
```diff
@@ -53,14 +55,62 @@ class DeconvolutionND(link.Link):
...
+
+    def __init__(self, ndim, in_channels, out_channels, ksize=None, stride=1,
+                 pad=0, nobias=False, outsize=None, initialW=None,
+                 initial_bias=None, dilate=1, groups=1):
         super(DeconvolutionND, self).__init__()
 
-        ksize = conv_nd.as_tuple(ksize, ndim)
+        if ksize is None:
+            ndim, out_channels, ksize, in_channels = \
+                ndim, in_channels, out_channels, None
```
Comment (hvy) [inline chainer/links/connection/deconvolution_nd.py]: Ditto regarding reassignment.

**Comment by hvy** [comment]
LGTM except minor comments.

File: chainer/links/connection/convolution_nd.py (hunk 5 above)
Comment (mitmul) [reply chainer/links/connection/convolution_nd.py]: Ooops, right

**Comment by mitmul** [comment]
@hvy Thanks, I fixed the points.
Jenkins, test this please

**Comment by mitmul** [comment]
jenkins test this please

**Comment by hvy** [comment]
LGTM!

**Pull request merged**